java/ql/lib/semmle/code/java/regex/regex.qll

```diff
@@ -0,0 +1,943 @@
+import java
+private import RegexFlowConfigs
+
+/**
+ * A string literal that is used as a regular exprssion.
+ */
+abstract class RegexString extends Expr {
+  RegexString() { this instanceof StringLiteral }
+
+  /** Holds if a character set starts between `start` and `end`. */
+  private predicate char_set_start0(int start, int end) {
+    this.nonEscapedCharAt(start) = "[" and
+    (
+      this.getChar(start + 1) = "^" and end = start + 2
+      or
+      not this.getChar(start + 1) = "^" and end = start + 1
+    )
+  }
+
+  /** Holds if the character at `pos` marks the end of a character class. */
+  private predicate char_set_end0(int pos) {
+    this.nonEscapedCharAt(pos) = "]" and
+    /* special case: `[]]` and `[^]]` are valid char classes.  */
+    not this.char_set_start0(_, pos - 1)
+  }
+
+  /**
+   * Gets the nesting depth of charcter classes at position `pos`
+   */
+  private int char_set_depth(int pos) {
+    exists(this.getChar(pos)) and
+    result =
+      max(int j |
+        j = 0 or
+        j =
+          count(int i | i < pos and this.char_set_start0(i, _)) -
```

Comment:
This is a question for people with better QL wizardry than me, but I think this will build intermediate expressions quadratic in the number of `[` or `]` characters in a string, which could be a problem if we run against a database with a very large string in it.

If that is a problem, a recursive version could work nicely, counting and maintaining state as it traverses a string from left to right.

Reply:
@aschackmull?